### PR TITLE
[bugfix] Resolve schedule step bug for PyTorch Profiler

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #<issue_number>

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/profiler/pytorch.py [hunk 1]
```diff
@@ -142,8 +145,9 @@ def _step(self) -> None:
         if self._current_action == "training_step_and_backward":
             self._num_training_step_and_backward += 1
         elif self._current_action == "validation_step":
-            if self._start_action_name == "on_train_start" and self._num_training_step_and_backward > 0:
-                self._num_validation_step += 1
+            if self._start_action_name == "on_fit_start":
+                if self._num_training_step_and_backward > 0:
```
Comment (SeanNaren) [inline pytorch_lightning/profiler/pytorch.py:149]: Unrelated to this PR and just my own curiosity, could you explain why we need to checkout `if self._num_training_step_and_backward > 0:`

Comment (tchaton) [reply pytorch_lightning/profiler/pytorch.py:149]: Because we don't want to record during sanity_checking.

**Review by SeanNaren — APPROVED** [review]
Not sure on all the technical details and you're much more versed, but looks reasonable to me

File: CHANGELOG.md [hunk 2]
```diff
@@ -191,6 +191,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed a bug where gradients were disabled after calling `Trainer.predict` ([#6657](https://github.com/PyTorchLightning/pytorch-lightning/pull/6657))
 
 
+- Fixed a bug with PyTorch Profiler ([#6674](https://github.com/PyTorchLightning/pytorch-lightning/pull/6674))
+
```
Comment (carmocca) [inline CHANGELOG.md:195]: Don't need this as the current changes are unreleased

```suggestion
```

File: pytorch_lightning/profiler/pytorch.py [hunk 3]
```diff
@@ -450,6 +451,9 @@ def summary(self) -> str:
 
         self._delete_profilers()
 
+        if self._schedule is not None:
+            self._schedule.reset()
+
```
Comment (carmocca) [inline pytorch_lightning/profiler/pytorch.py:456]: Should this be in `_delete_profilers` instead?

File: pytorch_lightning/profiler/pytorch.py [hunk 4]
```diff
@@ -422,12 +420,12 @@ def stop(self, action_name: str) -> None:
...
 
-        if action_name in self.step_action_names:
+        if self.profiler is not None and action_name in self.STEP_FUNCTIONS:
             if self._schedule is not None:
-                self._schedule._current_action = action_name
+                self._schedule.pre_step(action_name)
 
             def on_trace_ready(profiler):
-                filename = f"{action_name}_{self.local_rank}"
+                filename = f"{self._stage}_{action_name}_{self.local_rank}"
```
Comment (carmocca) [inline pytorch_lightning/profiler/pytorch.py:428]: you can remove this

```suggestion
```

and pass below:

```python
handler = tensorboard_trace_handler(self.dirpath, self._prepare_filename(extension=""))
```

File: pytorch_lightning/profiler/pytorch.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/profiler/pytorch.py:456]: I hesitated, but yes why not.

File: pytorch_lightning/profiler/pytorch.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/profiler/pytorch.py:428]: Done !

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/profiler/pytorch.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/profiler/pytorch.py:149]: Could you add this as a comment in the PR?

**Review by ananthsub — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/profiler/pytorch.py [hunk 5]
```diff
@@ -107,6 +107,15 @@ def __init__(self, schedule: Callable) -> None:
         if not _KINETO_AVAILABLE:
             raise ModuleNotFoundError("You are trying to use `ScheduleWrapper` which require kineto install.")
         self._schedule = schedule
+        self.reset()
```
Comment (Borda) [inline pytorch_lightning/profiler/pytorch.py:110]: this is the only diff here, right?
the rest for `setup` and `pre_step` seems the same

File: pytorch_lightning/profiler/pytorch.py [hunk 6]
```diff
@@ -374,10 +379,6 @@ def _default_activities(self) -> List['ProfilerActivity']:
             activities.append(ProfilerActivity.CUDA)
         return activities
 
-    @property
-    def step_action_names(self) -> Set[str]:
```
Comment (Borda) [inline pytorch_lightning/profiler/pytorch.py:378]: shall this be properly deprecated?

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/profiler/pytorch.py (hunk 6 above)
Comment (Borda) [reply pytorch_lightning/profiler/pytorch.py:378]: @tchaton ^^

Comment (carmocca) [reply pytorch_lightning/profiler/pytorch.py:378]: No, that function has not been released yet

**Pull request merged**